File: 2020/php/src/Aoc/Days/Day12.php
<?php


namespace Ppx17\Aoc2020\Aoc\Days;


use Illuminate\Support\Collection;
use Ppx17\Aoc2020\Aoc\Days\Common\Vector2;

class Day12 extends AbstractDay
{
    private Collection $instructions;
    private array $windDirections;

    public function dayNumber(): int
    {
        return 12;
    }

    public function setUp(): void
    {
        $this->instructions = collect($this->getInputLines())
            ->map(fn($x) => [$x[0], (int)substr($x, 1)]);

        $this->windDirections = [
            'E' => Vector2::east(),
            'S' => Vector2::south(),
            'W' => Vector2::west(),
            'N' => Vector2::north()
        ];
    }

    public function part1(): string
    {
        return $this->sail(new Vector2(1, 0), false);
    }

    public function part2(): string
    {
        return $this->sail(new Vector2(10, -1), true);
    }

    private function sail(Vector2 $direction, bool $moveDirection): int
    {
        $position = new Vector2();
        foreach ($this->instructions as $instruction) {
            switch ($instruction[0]) {
                case 'L':
                    $direction = $direction->rotateLeftTimes($instruction[1] / 90);
                    break;
                case 'R':
                    $direction = $direction->rotateRightTimes($instruction[1] / 90);
                    break;
                case 'N':
                case 'E':
                case 'W':
                case 'S':
                    if ($moveDirection) {
                        $direction = $direction->move($this->windDirections[$instruction[0]], $instruction[1]);
                    } else {
                        $position = $position->move($this->windDirections[$instruction[0]], $instruction[1]);
                    }
                    break;
                case 'F':
                    $position = $position->move($direction, $instruction[1]);
                    break;
            }
        }

        return $position->manhattan();
    }
}

File: 2020/julia/src/Day12.jl
module Day12
include("Aoc.jl")
using .Aoc

struct Instruction
    i::Char
    t::UInt32
end

instructions = [Instruction(x[1], parse(UInt32, x[2:end])) for x in Aoc.input_lines(12)]
wind_directions = Dict('E' => [0, 1], 'W' => [0, -1], 'N' => [-1, 0], 'S' => [1, 0])

rotR(vec::Array{Int64, 1}) = [vec[2], -vec[1]]

function sail(direction, move_dir)
    position = [0, 0]
    for i in instructions
        if i.i == 'L' || i.i == 'R'
            for _ in 1:((i.i == 'L' ? 360 - i.t : i.t) ÷ 90)
                direction = rotR(direction)
            end
        elseif i.i == 'F'
            position += direction * i.t
        else
            move_dir && (direction += wind_directions[i.i] * i.t)
            !move_dir && (position += wind_directions[i.i] * i.t)
        end
    end
    sum(abs.(position))
end

part1() = sail([0, 1], false)
part2() = sail([-1, 10], true)
end